lib/src/sql/statements/ifelse.rs

```diff
@@ -48,21 +50,21 @@ impl IfelseStatement {
 	}
 }
 
-impl fmt::Display for IfelseStatement {
+impl Display for IfelseStatement {
 	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		write!(
+		Display::fmt(
+			&Fmt::new(
+				self.exprs.iter().map(|args| {
+					Fmt::new(args, |(cond, then), f| write!(f, "IF {} THEN {}", cond, then))
+				}),
+				fmt_separated_by(" ELSE "),
+			),
 			f,
-			"{}",
-			self.exprs
-				.iter()
-				.map(|(ref cond, ref then)| format!("IF {} THEN {}", cond, then))
-				.collect::<Vec<_>>()
-				.join(" ELSE ")
 		)?;
 		if let Some(ref v) = self.close {
 			write!(f, " ELSE {}", v)?
 		}
-		write!(f, " END")?;
+		f.write_str(" END")?;
 		Ok(())
 	}
 }
```

Comment:
From now on, using `format!` in `impl Display` is not only inefficient but incorrect. Instead of passing through formatting actions, allowing the pretty printer to add newlines and indentation in between them, it buffers up formatting actions into a `String`.

The solution is to use `sql::fmt::Fmt` along side a formatting function. Since a common formatting activity is joining an iterator of `impl Display` by some other `impl Display`, I made a helper called `fmt_separated_by` that produces a closure capable of doing that. It's like a general version of `Fmt::[pretty_]comma_separated`.